src/main.ts

```diff
@@ -7,479 +7,1026 @@ import {
+    const keyPress = KeyPress.fromEvent(event);
+    const registerState = this.registerMachine.advance(keyPress);
+    writeConsole(
+      `An keypress resulted in ${MappingState[registerState]} state.`,
+    );
+
+    switch (registerState) {
+      case MappingState.EmptySequence:
+      case MappingState.WaitingInput:
+      case MappingState.FirstKey:
+      case MappingState.DeletedKey:
+      case MappingState.AddedKeys:
+        this.renderKeyPresses(this.registerMachine.documentRepresentation());
+        return;
+
+      case MappingState.PendingDeletion:
+      case MappingState.PendingAddition:
+        {
+          // Inplace mutation :(
+          const elements = this.registerMachine.documentRepresentation();
+          const lastElement = elements[elements.length - 1];
+          lastElement.style.opacity = '0.5';
+          this.renderKeyPresses(elements);
+
+          const backspace = KeyPress.just('Backspace').kbd();
+          const enter = KeyPress.just('Enter').kbd();
+          const ctrlAltEnter = KeyPress.ctrlAlt('Enter').kbd();
+          const pressLiteral = lastElement.cloneNode(true) as HTMLElement;
+          pressLiteral.style.opacity = '1';
+
+          const discardOrRemoveWith =
+            registerState === MappingState.PendingAddition
+              ? 'If not, discard it with '
+              : 'You can remove previous mapping it with ';
+
+          const confirmText = document.createElement('p');
+          confirmText.append(
+            'Did you mean literal ',
+            pressLiteral,
+            '?',
+            document.createElement('br'),
+            'If so, press ',
+            enter,
+            '.',
+            document.createElement('br'),
+            discardOrRemoveWith,
+            backspace,
+            '.',
+            document.createElement('br'),
+            'Press ',
+            ctrlAltEnter,
+            ' to discard pending changes and complete.',
+          );
+
+          this.contentEl.append(confirmText);
+        }
+        return;
+
+      case MappingState.FinishedMapping:
+        const keyPresses = [...this.registerMachine.presses()];
+        const conflicts = this.parent.conflicts(keyPresses);
+        if (conflicts.length >= 1) {
+          // todo handle this properly
+          createNotice('There are conflicts with your keyPresses!');
+        } else {
+          const newKeyMap = new KeyMap(this.commandId, keyPresses);
+          this.parent.addKeymap(newKeyMap);
+          const sequenceRepr = newKeyMap.sequence
+            .map((key) => key.text())
+            .join(' => ');
+          createNotice(`Command  ${this.commandId}
+           can now be invoked by ${sequenceRepr}`);
+          this.close();
+        }
     }
+  };
+
+  private readonly renderKeyPresses = (elements: HTMLElement[]): void => {
+    this.contentEl.empty();
+
+    const header = document.createElement('p');
+    header.setText(`Registering for ${this.commandId}`);
```

Comment:
It would be nice to have an informative title above this. It was unclear to me at first what was expected of me here.

It would also be nice to be able to finish the command sequence by clicking a "Save" button or something..

Reply:
A informative would be nice here. I tend to skimp on UI , sorry about that haha
Do you think a Save button is necessary? Were you able to complete the command using the 'enter' key, and following the suggestions?

Reply:
It took me a while to figure out that pressing a special key followed by another key combo would save. I think adding a Save and Cancel pair of buttons would definitely help make the UI easier to use. Leaving the key combo too is great though, there can be more than one way to save it.

Reply:
I've added the save button, will add the cancel too, nice heads up.